### PR TITLE
Configurable timeouts and suspended fixes

### PR DESCRIPTION
**Pull request by SimonWoolf**
Implementation of https://github.com/ably/docs/pull/51, as well as some fixes to make the `suspended` state work. (Not implemented yet pending discussion: `httpMaxRetryDuration`).

Not for merging until https://github.com/ably/docs/pull/51 is finalised and merged.

Closes https://github.com/ably/ably-js/issues/99

### CODE REVIEW

**Comment by SimonWoolf** [comment]
@paddybyers updated per final version of ably/docs#51; ready for reviewing.

Per [discussion](https://github.com/ably/docs/pull/51/files#r41623379), did not implement `httpOpenTimeout`.

Also did not implement `httpMaxRetryDuration`: ISTM its usefulness relies on `httpOpenTimeout`, since without that, if an http request is timing out rather than failing, it'll take `httpRequestTimeout`, which (by default, anyway) is longer than `httpMaxRetryDuration`, so its effect would be to not try any fallbacks at all if the first attempt times out, which doesn't seem ideal. WDYT?

**Comment by paddybyers** [comment]
I thought we had agreed that `httpRequestTimeout` and arealtimeRequestTimeout` are the same thing and should be 10s.

**Comment by mattheworiordan** [comment]
> I thought we had agreed that httpRequestTimeout and a realtimeRequestTimeout` are the same thing and should be 10s.

@paddybyers no, I don't think we agreed that.  We agreed to have one timeout that is used for all realtime operations, but that had no correlation with the REST based timeouts. See https://github.com/ably/docs/commit/2dc7923302d68f68aa1736dfe6497013c8725350

**Comment by SimonWoolf** [comment]
@paddybyers any more thoughts on this thread? if we're going to be doing more tweaks to the timeouts & fallbacks would be nice to have this merged

**Comment by paddybyers** [comment]
Lets merge it. It looks like it needs rebasing first.

**Comment by mattheworiordan** [comment]
Is this going to be merged?

File: common/lib/util/defaults.js [hunk 1]
```diff
@@ -5,12 +5,19 @@ Defaults.WS_HOST                  = 'realtime.ably.io';
...
-Defaults.disconnectTimeout        = 30000;
-Defaults.suspendedTimeout         = 120000;
-Defaults.recvTimeout              = 90000;
-Defaults.sendTimeout              = 10000;
-Defaults.connectionPersistTimeout = 15000;
+Defaults.TIMEOUTS = {
+	/* Documented as options params: */
+	disconnectedRetryTimeout : 15000,
+	suspendedRetryTimeout    : 30000,
+	httpRequestTimeout         : 15000,
```
Comment (mattheworiordan) [inline common/lib/util/defaults.js]: Odd alignment here

File: common/lib/util/defaults.js [hunk 2]
```diff
@@ -28,9 +35,10 @@ Defaults.getPort = function(options, tls) {
 
 Defaults.getHosts = function(options) {
 	var hosts = [options.host],
-		fallbackHosts = options.fallbackHosts;
+		fallbackHosts = options.fallbackHosts,
+		httpMaxRetryCount = typeof(options.httpMaxRetryCount) !== 'undefined' ? options.httpMaxRetryCount : Defaults.httpMaxRetryCount;
 
-	if(fallbackHosts) hosts = hosts.concat(fallbackHosts);
+	if(fallbackHosts) hosts = hosts.concat(fallbackHosts.slice(0, httpMaxRetryCount));
```
Comment (mattheworiordan) [inline common/lib/util/defaults.js]: I am not sure this is correct unless the fallback hosts have been randomised first

File: spec/browser/connection.test.js [hunk 3]
```diff
@@ -115,14 +109,13 @@ define(['ably', 'shared_helper'], function(Ably, helper) {
 			test.ok(false, 'connection to server failed');
 			closeAndFinish(test, realtime);
 		});
+		connection.on(function(){console.log(this.event)})
```
Comment (mattheworiordan) [inline spec/browser/connection.test.js]: Did you mean to commit this?

File: spec/rest/defaults.test.js [hunk 4]
```diff
@@ -36,7 +36,7 @@ define(['ably', 'shared_helper'], function(Ably, helper) {
 		test.deepEqual(normalisedOptions.fallbackHosts, Defaults.FALLBACK_HOSTS);
 		test.equal(normalisedOptions.tls, true);
 
-		test.deepEqual(Defaults.getHosts(normalisedOptions), [normalisedOptions.host].concat(Defaults.FALLBACK_HOSTS));
+		test.deepEqual(Defaults.getHosts(normalisedOptions), [normalisedOptions.host].concat(Defaults.FALLBACK_HOSTS.slice(0,3)));
```
Comment (mattheworiordan) [inline spec/rest/defaults.test.js]: Again, simply slicing effectively means we only ever use `a-c.ably-realtime.com` which is not the intention

**Comment by SimonWoolf** [comment]
@mattheworiordan this was merged a couple weeks ago as https://github.com/ably/ably-js/commit/8537720b755e8c95dd606258052c69dc951259d7. Sorry, forgot to close the PR.

**Comment by SimonWoolf** [comment]
(This was an intermediate PR; new issue for discussing timeouts is at https://github.com/ably/ably-js/issues/165)

**Pull request closed without merging**